### PR TITLE
Add FSDP docs

### PR DESCRIPTION
**Pull request by SeanNaren**
## What does this PR do?

Adds docs for #7487, with usage instructions for this V1 case

cc @shuyingsunshine21 @ananthsub 

## Before submitting
- [ ] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: docs/source/advanced/advanced_gpu.rst [hunk 1]
```diff
@@ -73,6 +73,103 @@ Sharded Training can work across all DDP variants by adding the additional ``--p
...
 Internally we re-initialize your optimizers and shard them across your machines and processes. We handle all communication using PyTorch distributed, so no code changes are required.
 
+----------
+
+.. _fully-sharded:
+
+Fully Sharded Training
+^^^^^^^^^^^^^^^^^^^^^^
+
+.. note::
```
Comment (justusschock) [inline docs/source/advanced/advanced_gpu.rst:83]: maybe make this a warning instead of a note? I think this is not a note as it does not state anything about the current behaviour

File: docs/source/advanced/advanced_gpu.rst [hunk 2]
```diff
@@ -73,6 +73,103 @@ Sharded Training can work across all DDP variants by adding the additional ``--p
...
+Wrap the Model
+""""""""""""""
+
+To activate parameter sharding, you must wrap your model using provided ``wrap`` or ``auto_wrap`` functions as described below. Internally in Lightning, we enable a context manager around the ``configure_sharded_model`` function to make sure the ``wrap`` and ``auto_wrap`` parameters are passed correctly.
+
+When not using Fully Sharded these wrap functions are a no-op. This means once the changes have been made, there is no need to remove the changes for other plugins.
+
+This is a requirement for really large models and also saves on instantiation time as modules are sharded instantly, rather than after the entire model is created in memory.
+
+``auto_wrap`` will recursively wrap modules within the ``LightningModule`` with nested Fully Sharded Wrappers,
```
Comment (justusschock) [inline docs/source/advanced/advanced_gpu.rst:109]: maybe state here that it wraps instances of `torch.nn.Module`? Otherwise a module can be arbitrary...

File: docs/source/advanced/advanced_gpu.rst [hunk 3]
```diff
@@ -73,6 +73,103 @@ Sharded Training can work across all DDP variants by adding the additional ``--p
...
+
+.. note::
+    Currently Fully Sharded Training relies on the user to wrap the model with Fully Sharded within the ``LightningModule``.
+    This means you must create a single model that is treated as a ``nn.Module`` within the ``LightningModule``.
+    This is a limitation of Fully Sharded Training that will be resolved in the future.
+
+Wrap the Model
+""""""""""""""
+
+To activate parameter sharding, you must wrap your model using provided ``wrap`` or ``auto_wrap`` functions as described below. Internally in Lightning, we enable a context manager around the ``configure_sharded_model`` function to make sure the ``wrap`` and ``auto_wrap`` parameters are passed correctly.
```
Comment (justusschock) [inline docs/source/advanced/advanced_gpu.rst:103]: Not really docs, but came to my mind (ignore it if not appropriate): Is it possible to add a maximum recursion depth here as a stopping condition for recursion instead of just the parameter number?

**Review by justusschock — APPROVED** [review]

**Review by awaelchli — APPROVED** [review]

**Review by Borda — APPROVED** [review]

**Review by ethanwharris — APPROVED** [review]

**Comment by SeanNaren** [comment]
Test is failing atm, but let's keep it like that cos I want @ananthsub and @shuyingsunshine21 to take a look :)

File: docs/source/advanced/advanced_gpu.rst [hunk 4]
```diff
@@ -73,6 +73,106 @@ Sharded Training can work across all DDP variants by adding the additional ``--p
...
+^^^^^^^^^^^^^^^^^^^^^^
+
+.. warning::
+    Fully Sharded Training is in beta and the API is subject to change. Please create an `issue <https://github.com/PyTorchLightning/pytorch-lightning/issues>`_ if you run into any issues.
+
+`Fully Sharded <https://fairscale.readthedocs.io/en/latest/api/nn/fsdp.html>`__ shards optimizer state, gradients and parameters across data parallel workers. This allows you to fit much larger models onto multiple GPUs into memory.
+
+Fully Sharded Training alleviates the need to worry about balancing layers onto specific devices using some form of pipe parallelism, and optimizes for distributed communication with minimal effort.
+
+By default, Fully Sharded acts similar to :ref:`sharded` which shards optimizer states and gradients. If you can train with default Fully Sharded, it is recommended to just use :ref:`sharded`.
```
Comment (shuyingsunshine21) [inline docs/source/advanced/advanced_gpu.rst:90]: @SeanNaren , which arg is controlling FSDP to be just sharing optimizer states and gradients?

**Comment by shuyingsunshine21** [comment]
lgtm overall, thanks @SeanNaren

File: docs/source/advanced/advanced_gpu.rst (hunk 4 above)
Comment (SeanNaren) [reply docs/source/advanced/advanced_gpu.rst:90]: Ah this is a mistake, by default we do shard activations as well, but if we set `reshard_after_forward=False` then we'll use ZeRO 2 (gradients/optimizers only) from what I've gathered in the [FSDP](https://fairscale.readthedocs.io/en/stable/api/nn/fsdp_tips.html) docs

**Comment by kaushikb11** [comment]
Re-running the jobs!

File: docs/source/advanced/advanced_gpu.rst (hunk 4 above)
Comment (shuyingsunshine21) [reply docs/source/advanced/advanced_gpu.rst:90]: got it. let us fix this then.
